src/modules/wallet/WalletWatcherIO.ts

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************************
+
+     The functions to provide real-time events to wallet clients are included.
+     The type of real-time event is when a block is created and
+     when a transaction of a specific address occurs.
+
+    Copyright:
+        Copyright (c) 2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+import { Height } from "boa-sdk-ts";
+
+import { Socket } from "socket.io";
+
+/**
+ * A class that optionally delivers real-time events to only one client who subscribed.
+ * The type of real-time event is when a block is created and
+ * when a transaction of a specific address occurs.
+ */
+export class WalletWatcherIO {
+    private tables: Map<string, Socket[]> = new Map<string, Socket[]>();
+    private ping: Map<string, number> = new Map<string, number>();
+
+    public onClientConnected(client_socket: Socket) {
+        this.ping.set(client_socket.id, new Date().getTime());
+
+        client_socket.on("subscribe", (data: { address: string }) => {
+            this.ping.set(client_socket.id, new Date().getTime());
+            const values = this.tables.get(data.address);
+            if (values === undefined) this.tables.set(data.address, [client_socket]);
+            else {
+                if (values.find((m) => m === client_socket) === undefined) values.push(client_socket);
+            }
+        });
+
+        client_socket.on("unsubscribe", (data: { address: string }) => {
+            this.ping.set(client_socket.id, new Date().getTime());
+            const values = this.tables.get(data.address);
+            if (values !== undefined) {
+                const found = values.findIndex((m) => m === client_socket);
+                if (found >= 0) values.splice(found, 1);
+            }
+        });
+
+        client_socket.on("ping", (data: any) => {
+            this.ping.set(client_socket.id, new Date().getTime());
+        });
+
+        client_socket.on("disconnect", () => {
```

Comment:
If the client shuts down abnormally without calling disconnet, garbage sockets may accumulate.

Reply:
I added the following functions.
The client sends a ping to the server once every 30 seconds.
The server calls the function checkConnection() every minute.
This function disconnects a client whose ping has been latest received for more than one minute.
And remove it from the subscription table.